server/src/main/java/org/elasticsearch/snapshots/SnapshotShutdownProgressTracker.java

```diff
@@ -0,0 +1,221 @@
+                case PAUSED -> pausedCount.incrementAndGet();
+                // The other stages are active, we should only see the end result because this method is called upon completion.
+                default -> {
+                    assert false : "unexpected shard snapshot stage: " + stage;
+                }
+            }
+        }
+    }
+
+    /**
+     * Uniquely tracks a request to update a shard snapshot status sent to the master node. Idempotent, safe to call multiple times.
+     *
+     * @param snapshot first part of a unique identifier
+     * @param shardId second part of a unique identifier
+     */
+    public void trackRequestSentToMaster(Snapshot snapshot, ShardId shardId) {
+        shardSnapshotRequests.add(snapshot.toString() + shardId.getIndexName() + shardId.getId());
+    }
+
+    /**
+     * Stops tracking a request to update a shard snapshot status sent to the master node. Idempotent, safe to call multiple times.
+     *
+     * @param snapshot first part of a unique identifier
+     * @param shardId second part of a unique identifier
+     */
+    public void releaseRequestSentToMaster(Snapshot snapshot, ShardId shardId) {
+        shardSnapshotRequests.remove(snapshot.toString() + shardId.getIndexName() + shardId.getId());
+    }
+
+    /**
+     * Logs some statistics about shard snapshot progress, as well as {@link HotThreads#logLocalHotThreads} to surface anything unexpected.
+     */
+    private void logProgressReport() {
+        // A shard snapshot has two phases for tracking, for reporting purposes. The first is while the shard snapshot is running on the
+        // data node. The second, to update the shard snapshot status in the cluster state on the master node, occurs asynchronously upon
+        // shard snapshot completion.
+        logger.info(
+            """
+                Current active shard snapshot stats. \
+                Node shutdown cluster state update received at [{}}]. \
+                Finished signalling shard snapshots to pause [{}]. \
+                [Phase 1 of 2] Number shard snapshots running on the data node (have not yet noticed pause signal) [{}]. \
+                [Phase 2 of 2] Number shard snapshots running according to the master node (awaiting response to status update \
+                requests sent to master node) [{}]. \
+                Shard snapshot completion stats during shutdown: Done [{}]; Failed [{}]; Aborted [{}]; Paused [{}] \
+                """,
+            shutdownStartMillis,
+            shutdownFinishedSignallingPausingMillis,
+            numberOfShardSnapshotsInProgressOnDataNode.get(),
+            shardSnapshotRequests.size(),
+            doneCount.get(),
+            failureCount.get(),
+            abortedCount.get(),
+            pausedCount.get()
+        );
+
+        HotThreads.logLocalHotThreads(logger, Level.INFO, "slow snapshot pausing", ReferenceDocs.STATELESS_EMPTY);
```

Comment:
On further thought, I don't think this belongs in the tracker. I'll pull it out and have it as a standalone someplace for dumping hot threads during slow shutdown, unassociated with snapshots in particular.